File: binary_search/python/binary_search.py
#! /usr/bin/ python3
import time
import sys

args = sys.argv
db =  list(range(1, int(args[1]) + 1))

def binary_search(data, item):
    """
    O(log n)
    """
    st = time.time()
    low = 0
    high = len(data) - 1
    
    while low <= high:
        mid = (low + high)  // 2
        guess = data[mid]

        if guess == item:
            et = time.time()
            print(f"Result: {mid}")
            return f"Binary Search Exec: {(et - st) * 1000}ms"
        if guess > item:
            high = mid - 1
        else:
            low = mid + 1




def linear_search(data, item):
    """
    O(n)
    """
    st = time.time()
    for i in data:
        if data[i] == item:
            et = time.time()
            print(f"Result: {data[i] - 1}")
            return f"Linear Search Exec: {(et - st) * 1000}ms"
        else:
            i += 1

print(binary_search(db, int(args[1])))
print("--------------------")
print(linear_search(db, int(args[1])))

File: binary_search/js/binary_search.js
function createArray(length) {
    const array = new Array(length)

    for (var i = 0; i < length; i++) {
        array[i] = i + 1; //This populates the array.  +1 is necessary because arrays are 0 index based and you want to store 1-100 in it, NOT 0-99.
    }
    return array
}

function binary_search(data, item) {
    let low = 0
    let high = data.length - 1
    let st = performance.now()

    while (low <= high) {
        let mid = Math.floor((low + high) / 2)
        let guess = data[mid]

        if (guess == item) {
            let et = performance.now()
            console.log(`Result: ${mid}`)
            console.log(`Binary Search Exec: ${et - st}ms`)
            return
        }
        if (guess > item) {
            high = mid - 1
        }
        else {
            low = mid + 1
        }
    }
}

function linear_search(data, item) {
    let st = performance.now()
    for (let i = 0; i < data.length; i++) {
        if (data[i] == item) {
            let et = performance.now()
            console.log(`Result: ${data[i] - 1}`)
            console.log(`Linear Search Exec: ${ et - st}ms`)
        }
    }
}

const args = process.argv

if (args[2]) {
    binary_search(createArray(args[2]), args[2])
    console.warn("---------------------")
    linear_search(createArray(args[2]), args[2])
} else {
    console.log("No argument provided, please enter a number")
}
